Nonnegative lars

michigraber opened the pull request:
I extended the lars_path, Lars and LarsLasso estimators in the scikit-learn least_angle.py module with the possibility to restrict coefficients to be >= 0 using the method described in the original paper by Efron et al, 2004, chapter 3.4.

michigraber committed:
lars_path method complemented with nonnegative option.

michigraber committed:
lars_path method complemented with nonnegative option.

michigraber committed:
lars_path method complemented with nonnegative option.

michigraber committed:
Merge branch 'nonnegative-lars' of https://github.com/michigraber/scikit-learn into nonnegative-lars

michigraber committed:
"nonnegative" -> "positive" + cleanup

michigraber committed:
"nonnegative" -> "positive" + cleanup

michigraber committed:
Merge branch 'nonnegative-lars' of https://github.com/michigraber/scikit-learn into nonnegative-lars

michigraber committed:
cleanup

michigraber committed:
todos for pull request

michigraber committed:
docstrings added + fit_intercept reintroduced.

michigraber committed:
cleanup

michigraber commented:
I realize it's necessary to extend the positivity parameter to the other estimators in this module as well ..

michigraber committed:
positivity constrained can now also be used for the CV estimators.

michigraber committed:
doctest fix.

michigraber committed:
positive option for Lars propagated to doc.

jnothman commented:
Please add tests comparable to https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/tests/test_coordinate_descent.py#L179

amueller commented:
Thanks. I'm sure @agramfort is interested (and can review ;)

michigraber commented:
will do so!
mi
On Aug 18, 2015 5:56 PM, "jnothman" <notifications@github.com> wrote:

> Please add tests comparable to
> https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/tests/test_coordinate_descent.py#L179
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/5131#issuecomment-132124704>
> .
>

agramfort commented on sklearn/linear_model/least_angle.py:
```diff
@@ -235,7 +247,10 @@ def lars_path(X, y, Xy=None, Gram=None, max_iter=500,
 
-            sign_active[n_active] = np.sign(C_)
+            if positive:
+                sign_active[n_active] = np.ones(C_.shape)
```
np.ones(C_.shape) -> np.ones_like(C_)

agramfort commented on sklearn/linear_model/least_angle.py:
```diff
@@ -354,7 +369,10 @@ def lars_path(X, y, Xy=None, Gram=None, max_iter=500,
 
         g1 = arrayfuncs.min_pos((C - Cov) / (AA - corr_eq_dir + tiny))
         g2 = arrayfuncs.min_pos((C + Cov) / (AA + corr_eq_dir + tiny))
```
you don't need g2 if positive

agramfort commented on sklearn/linear_model/least_angle.py:
```diff
@@ -978,8 +1018,9 @@ class LarsCV(Lars):
                  normalize=True, precompute='auto', cv=None,
                  max_n_alphas=1000, n_jobs=1, eps=np.finfo(np.float).eps,
-                 copy_X=True):
+                 copy_X=True, positive=False):
```
position of the positive param is not consistent for both classes. I would put it always as last param to avoid breaking code.

agramfort commented on sklearn/linear_model/least_angle.py:
```diff
@@ -772,13 +801,15 @@ class LassoLars(Lars):
                  normalize=True, precompute='auto', max_iter=500,
-                 eps=np.finfo(np.float).eps, copy_X=True, fit_path=True):
+                 positive=False, eps=np.finfo(np.float).eps, copy_X=True,
+                 fit_path=True):
```
here positive is in the middle.

agramfort commented:
you need to add tests. See how the positive option is tested for Lasso class.

agramfort commented:
ping me when done. thanks @michigraber

michigraber committed:
agramfort code review input

michigraber commented:
thanks for the input at @agramfort ! it's ingested.

michigraber committed:
test lars_path positive constraint.

michigraber committed:
merge with upstream/master

michigraber committed:
fit_intercept docstring comment removed from lars_path method

michigraber committed:
positive option passed on to LarsCV estimator

michigraber committed:
tests of positive option for estimator classes

michigraber committed:
tests of positive option for estimator classes condensed.

michigraber commented:
@agramfort and @jnothman, I added tests to check for positivity of coefficients when using the positive option using the diabetes dataset (there are negative coefficients without the option, it's also tested) for the lars_path method and all estimator classes.

agramfort commented on sklearn/linear_model/least_angle.py:
```diff
@@ -772,13 +801,15 @@ class LassoLars(Lars):
         self.verbose = verbose
         self.normalize = normalize
         self.method = 'lasso'
+        self.positive=positive 
```
pep8 : space around =

agramfort commented on sklearn/linear_model/tests/test_least_angle.py:
```diff
@@ -439,4 +439,70 @@ def test_lars_path_readonly_data():
+        assert_true(estimator.coef_.min() < 0)
+        estimator = getattr(linear_model, estname)(positive=True, **params)
+        estimator.fit(diabetes['data'], diabetes['target'])
+        assert_true(min(estimator.coef_) >= 0)
```
how long does it take to run?

can you add a test that compares the resets of Lasso and LassoLars with positive == True? there is already a test. Just update it to also test with positive=True.

michigraber replied:
> In sklearn/linear_model/tests/test_least_angle.py:
> 
> > +        'LassoLarsIC' : { },
> > +        }
> > +
> > +def test_estimatorclasses_positive_constraint():
> > +    # testing the transmissibility for the positive option of all estimator
> > +    # classes in this same function here 
> > +
> > +    for estname in estimator_parameter_map:
> > +        params = default_parameter.copy()
> > +        params.update(estimator_parameter_map[estname])
> > +        estimator = getattr(linear_model, estname)(positive=False, **params)
> > +        estimator.fit(diabetes['data'], diabetes['target'])
> > +        assert_true(estimator.coef_.min() < 0)
> > +        estimator = getattr(linear_model, estname)(positive=True, **params)
> > +        estimator.fit(diabetes['data'], diabetes['target'])
> > +        assert_true(min(estimator.coef_) >= 0)
> how long does it take to run?
> 
In [2]: %timeit test_least_angle.test_estimatorclasses_positive_constraint()
10 loops, best of 3: 33 ms per loop

michigraber committed:
pep-8 fix

michigraber committed:
tests for comparison of results for Lasso and LarsLasso added.

michigraber commented:
@agramfort : About the test comparing Lasso and LarsLasso with the positive option:

I had to disentangle this one a bit because the tests do not trivially pass. So I added two more functions instead of simply extending the existing one.

Two parts validate without problems, they're in test_lasso_lars_vs_lasso_cd_positive().
The middle part however, where a range of alphas is used leads to too big coefficient differences.

I isolated this test in another function called evaluate_lasso_lars_vs_lasso_cd_positive_middle_part() to get an idea of why it fails. Here we have some printed output:

In [19]: test_least_angle.evaluate_lasso_lars_vs_lasso_cd_positive_middle_part()
alpha=0.01 	                 err=25.7243223179 	         num_coeffs=(5/5) (LassoLars / Lasso)
alpha=0.0615789473684 	 err=18.2545548874 	 num_coeffs=(5/5) (LassoLars / Lasso)
alpha=0.113157894737 	 err=10.7847873501 	         num_coeffs=(5/5) (LassoLars / Lasso)
alpha=0.164736842105 	 err=3.31501982266 	         num_coeffs=(5/5) (LassoLars / Lasso)
alpha=0.216315789474 	 err=3.69727628935e-06 	 num_coeffs=(4/4) (LassoLars / Lasso)
alpha=0.267894736842 	 err=3.56502617927e-06 	 num_coeffs=(4/4) (LassoLars / Lasso)
alpha=0.319473684211 	 err=3.43302342569e-06 	 num_coeffs=(4/4) (LassoLars / Lasso)
alpha=0.371052631579 	 err=1.05509699804e-06 	 num_coeffs=(3/3) (LassoLars / Lasso)
alpha=0.422631578947 	 err=5.9269674114e-07 	         num_coeffs=(3/3) (LassoLars / Lasso)
alpha=0.474210526316 	 err=1.56511823248e-06 	 num_coeffs=(3/3) (LassoLars / Lasso)
alpha=0.525789473684 	 err=1.28253471214e-06 	 num_coeffs=(3/3) (LassoLars / Lasso)
alpha=0.577368421053 	 err=1.00002954531e-06 	 num_coeffs=(3/3) (LassoLars / Lasso)
alpha=0.628947368421 	 err=7.38618104242e-07 	 num_coeffs=(3/3) (LassoLars / Lasso)
alpha=0.680526315789 	 err=6.78234533087e-07 	 num_coeffs=(3/3) (LassoLars / Lasso)
alpha=0.732105263158 	 err=2.25315273789e-06 	 num_coeffs=(3/3) (LassoLars / Lasso)
alpha=0.783684210526 	 err=1.98825668767e-06 	 num_coeffs=(3/3) (LassoLars / Lasso)
alpha=0.835263157895 	 err=1.72877180183e-06 	 num_coeffs=(3/3) (LassoLars / Lasso)
alpha=0.886842105263 	 err=1.58242049123e-06 	 num_coeffs=(3/3) (LassoLars / Lasso)
alpha=0.938421052632 	 err=1.50923193078e-06 	 num_coeffs=(3/3) (LassoLars / Lasso)
alpha=0.99 	                 err=1.43604831986e-06 	 num_coeffs=(3/3) (LassoLars / Lasso)

The number of regressors in the diabetes dataset btw is 10, but even for very small alphas not more than 5 are selected with the positive option in the case of both estimators.

We find that the comparison fails for small alphas. This is in congruence with a statement in the original publication of the least angle regression algorithm by Efron et al. 2004:

* The positive Lasso usually does not converge to the full OLS solution beta_{m}, even fro very large choices of t.*

This is kind of a 'flaw' of the method. For 'reasonable' alphas (ie also the default value alpha=1.0) the algorithm is of value ..

How shall we account for this?

agramfort commented:
did you look at the path with positive=True for Lasso and LassoLars using
for example

http://scikit-learn.org/stable/auto_examples/linear_model/plot_lasso_coordinate_descent_path.html

how much do they match on the grid of alphas?

michigraber commented:
I made a ipython notebook gist for this: https://gist.github.com/michigraber/7e7d7c75eca694c7a6ff

ogrisel commented:
> How shall we account for this? A comment in the docstring?

+1 for a comment in the docstring and an inline comment in the tests that explains why we don't check for small values of alpha.

michigraber committed:
docstrings updated with considerations regarding small alphas when positive=True.

michigraber committed:
comparison test of Lasso and LassoLars under positive restriction refactored and commented

michigraber commented:
@ogrisel and @agramfort : I added considerations about the usage of the positive option to the functions and estimators in question. Please let me know if you do not agree or understand them.

Also I updated the test function to now only compare a restricted range of alpha values and added some inline comments.

agramfort commented:
thanks @michigraber merging.

I'll push a pep8 cosmit in master

agramfort commented:
cosmit pushed.

next time avoid merging master on the way. It prevented me from doing a clean rebase.

agramfort commented:
for the record https://github.com/scikit-learn/scikit-learn/commit/a8585a4b0d32f84ac389e8eafa4a80b50aa8b7af

michigraber commented:
cool! thanks for supporting guys!
